src/core/lib/resource_quota/memory_quota.h

```diff
@@ -0,0 +1,336 @@
+class MemoryAllocator final : public InternallyRefCounted<MemoryAllocator> {
+ public:
+  explicit MemoryAllocator(RefCountedPtr<MemoryQuota> memory_quota);
+  ~MemoryAllocator() override;
+
+  void Orphan() override;
+
+  // Rebind -  Swaps the underlying quota for this allocator, taking care to
+  // make sure memory allocated is moved to allocations against the new quota.
+  void Rebind(RefCountedPtr<MemoryQuota> memory_quota)
+      ABSL_LOCKS_EXCLUDED(memory_quota_mu_);
+
+  // Reserve bytes from the quota.
+  // If we enter overcommit, reclamation will begin concurrently.
+  // Returns the number of bytes reserved.
+  size_t Reserve(MemoryRequest request) ABSL_LOCKS_EXCLUDED(memory_quota_mu_);
+
+  // Release some bytes that were previously reserved.
+  void Release(size_t n) ABSL_LOCKS_EXCLUDED(memory_quota_mu_) {
+    // Add the released memory to our free bytes counter... if this increases
+    // from  0 to non-zero, then we have more to do, otherwise, we're actually
+    // done.
+    if (free_bytes_.fetch_add(n, std::memory_order_release) != 0) return;
+    MaybeRegisterReclaimer();
+  }
+
+  // Post a reclaimer for some reclamation pass.
+  void PostReclaimer(ReclamationPass pass,
+                     std::function<void(ReclamationSweep)>);
+
+  // Allocate a new object of type T, with constructor arguments.
+  // The returned type is wrapped, and upon destruction the reserved memory
+  // will be released to the allocator automatically. As such, T must have a
+  // virtual destructor so we can insert the necessary hook.
+  template <typename T, typename... Args>
+  absl::enable_if_t<std::has_virtual_destructor<T>::value, T*> New(
+      Args&&... args) ABSL_LOCKS_EXCLUDED(memory_quota_mu_) {
+    // Wrap T such that when it's destroyed, we can release memory back to the
+    // allocator.
+    class Wrapper final : public T {
+     public:
+      explicit Wrapper(RefCountedPtr<MemoryAllocator> allocator, Args&&... args)
+          : T(std::forward<Args>(args)...), allocator_(std::move(allocator)) {}
+      ~Wrapper() override { allocator_->Release(sizeof(*this)); }
+
+     private:
+      const RefCountedPtr<MemoryAllocator> allocator_;
+    };
+    Reserve(sizeof(Wrapper));
+    return new Wrapper(Ref(), std::forward<Args>(args)...);
+  }
+
+  // Construct a unique ptr immediately.
+  template <typename T, typename... Args>
+  std::unique_ptr<T> MakeUnique(Args&&... args)
+      ABSL_LOCKS_EXCLUDED(memory_quota_mu_) {
+    return std::unique_ptr<T>(New<T>(std::forward<Args>(args)...));
+  }
+
+ private:
+  // Result of TryReserve.
+  // If memory was allocated, success = true and size = the number of bytes
+  // allocated. If memory was not allocated, success = false and size = the
+  // number of bytes to take from the quota.
+  struct ReserveResult {
+    bool success;
+    size_t size;
+  };
+
+  // Primitive reservation function.
+  ReserveResult TryReserve(MemoryRequest request) GRPC_MUST_USE_RESULT;
```

Comment:
nit: `absl::optional<size_t>` is pretty much identical to this.

Reply:
Yeah... it was fancier a few revisions ago... I think it's just about back to that though.